realtek: Interrupt handling fixes (v3)

plappermaul opened the pull request:
Continuation of #10562 after a messed up rebase.

Thinking about this patch and reading [this conversation](https://lore.kernel.org/all/763394a6e5c83006eb4628a9d0242b7eb04b889d.camel@svanheule.net/) I strongly believe that IRR is indeed a routing priority for the interrupts. With this my patch will read "use routing priority 0 to disable interrupts at all". If the "priority idea" makes sense then [something like this](https://elixir.bootlin.com/linux/v5.10.139/source/Documentation/devicetree/bindings/interrupt-controller/ti,c64x+megamod-pic.txt) might closely match our observation

hauke commented:
Please do not create a new pull request for changes to an existing pull request, but force push to your existing branch to update the existing pull request.

Borromini commented on target/linux/realtek/patches-5.10/320-irqchip-irq-realtek-use-hotplug.patch:
```diff
@@ -0,0 +1,85 @@
+ 	CPUHP_AP_IRQ_MIPS_GIC_STARTING,
+ 	CPUHP_AP_IRQ_RISCV_STARTING,
+ 	CPUHP_AP_IRQ_SIFIVE_PLIC_STARTING,
++	CPUHP_AP_IRQ_REALTEK_STARTING,
```
Shouldn't this be sorted alphabetically, before `CPUHP_AP_IRQ_RISCV_STARTING`?

plappermaul replied:
I already checked but the whole timers list is not sorted at all.

plappermaul commented:
Additional feedback. Testing with RLT839X devices there seems to be one glitch left. Additional fix needed:

```
        for_each_cpu(cpu, &cpu_disable) {
                write_irr(REG(RTL_ICTL_IRR0, cpu), i->hwirq, 0);
                realtek_ictl_unmask[cpu] &= ~BIT(i->hwirq);

+               value = readl(REG(RTL_ICTL_GIMR, cpu)); /* Disable GIMR immediately */
+               value &= ~BIT(i->hwirq);
+               writel(value, REG(RTL_ICTL_GIMR, cpu));
        }

        for_each_cpu(cpu, &cpu_enable) {
                write_irr(REG(RTL_ICTL_IRR0, cpu), i->hwirq, output->output_index + 1);
                realtek_ictl_unmask[cpu] |= BIT(i->hwirq);

+               value = readl(REG(RTL_ICTL_GIMR, cpu)); /* Enable GIMR immediately */
+               value |= (BIT(i->hwirq) & realtek_ictl_unmask[cpu]);
+               writel(value, REG(RTL_ICTL_GIMR, cpu));
        }
```

plappermaul committed:
realtek: avoid wrong interrupt routing

The interrupt controller depends on two control registers. GIMR enables
or disables interrupts and IRRx routes these to MIPS CPU interrupts 2-7.
Wiki currently states "A value of '0' (in IRRx) disconnects this input from
the output line, independent of the line's setting in GIMR."

Contrary to normal intuition this statement DOES NOT mean, that interrupts
can be disabled by IRRx alone. The sad truth was discovered by enabling
SMP for an Zyxel XGS1010 on the 930x target. It shows that driver and
interrupts behave as follows:

- Timer 0 interrupt 7 has active routing to CPU0 and no routing to CPU1
- Timer 1 interrupt 8 has no routing to CPU0 and active routing to CPU1
- Unmasking (enabling) interrupts writes 1 bits to all GIMR registers
- Masking (disabling) interrupts writes 0 bits to both GIMR registers

During operation we can encounter a situation like

- GIMR bit for a interrupt/CPU combination is set to enabed (=1)
- IRRx routing bits for a interrupt/CPU combination are set to disabed (=0)

This setting already allows the hardware to fire interrupts to the target
CPU/VPE if the other CPU/VPE is currently busy. Especially for CPU bound
timer interrupts this is lethal. If timer interrupt 7 arrives at CPU1 and
vice versa for interrupt 8 the restart trigger gets lost. The timer dies
and a msleep() operation in the kernel will halt endlessly.

Fix this by tracking the IRRx active routing setting in a new bitfield with
0="routing active" and 1="no routing". Enable interrupts in GIMR only
for a interrupt & CPU if routing is active. Thus we have

- GIMR = 0 / IRRx = 0 -> everything disabled
- GIMR = 1 / IRRx > 0 -> active and normal routing
- GIMR = 0 / IRRx > 0 -> masked (disabled) with normal routing
- GIMR = 1 / IRRx = 0 -> no longer possible

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul commented:
@svanheule after extensive testing with the new timer driver the patches should be fine now.

The latest version takes care about following situation:

affinity is changed
GIMR was not updated
next (timer) interrupt lands on the wrong CPU

abajk commented on target/linux/realtek/patches-5.10/320-irqchip-irq-realtek-use-hotplug.patch:
```diff
@@ -0,0 +1,85 @@
++ * the interrupts for the second CPU/VPE. Fix this during hotplug.
++ */
++	if (cpu > 0)
++		change_c0_status(ST0_IM,
```
This can be shortened:
`set_c0_status(ST0_IM);`

abajk commented:
FYI, there was a r[ecent discussion](https://lkml.org/lkml/2022/7/2/237) on how to enable interrupts for a second VPE on a 34Kc core. I will check if your solution also works on Lantiq SoCs.

plappermaul commented:
Thanks for pointing that out. 

In our case we had no problems with the interrupt enabling. Birger had patched the kernel already. We had routing issues with interrupts reaching wrong CPUs. 

Nevertheless if this patch will go mainline we can avoid fixing it during hotplug.

plappermaul commented:
@abajk while you were here ... 

Can you give a hint if a CPU mask/unmask operation must run for all VPEs or only the for one it is running on?  

In our implementation we do it always for all VPEs. But from reading other interrupt controllers I'm not sure if only a local mask/unmask might right.

svanheule commented:
If we don't mask on all VPEs, then both might try to service the interrupt. Not sure how much of an issue that might be though, or if the irqchip subsystem already takes care of that issue.

abajk commented:
On Lantiq, one interrupt is distributed to only one of the VPEs. I'm not entirely sure, but some interrupt controllers can distribute interrupts to multiple cores using the round robin algorithm. This is hardware dependent.

svanheule commented:
As far as I can tell from the register descriptions, the Realtek interrupt controller is duplicated per VPE. This allows for a peripheral to connect independent IRQ lines to both VPEs, which is used on RTL930x SoCs (and I suspect also some WiFi SoCs) to give IRQ affinity capabilites to the GPIO interrupt controller.

```
SOC IRQ LINES                        MIPS
                       +-----------+ HW IRQ   +--------+
Peripheral A[CPU0]---->| SOC IRQ   | LINES    | IRQ    |
Peripheral B ------o-->| VPE0 CTRL |=========>| VPE0   |
        .          |   +-----------+          +--------+
        .          |
        .          |
                   |   +-----------+          +--------+
                   \-->| SOC IRQ   |          | IRQ    |  (Not available ont RTL8380)
Peripheral A[CPU1]---->| VPE1 CTRL |=========>| VPE1   |
                       +-----------+          +--------+
```

plappermaul commented:
Corresponding Lantiq PR #10676

oliv3r commented:
With the new insights gained from this MR;
> The interrupt controller depends on two control registers. GIMR enables
or disables interrupts and IRRx routes these to MIPS CPU interrupts 2-7.

GIMR (obviously) Global Interupt Masking Register
IRRx (maybe now obviously) Interrupt Routing Register

I wonder if we should document all this gained knowledge where it can be more easily found (wiki?/kernel-readme) as we don't have a datasheet, all this knowledge evaporates in time again ...

oliv3r commented on target/linux/realtek/patches-5.10/319-irqchip-irq-realtek-rtl-fix-VPE-affinity.patch:
```diff
@@ -0,0 +1,145 @@
++++ b/drivers/irqchip/irq-realtek-rtl.c
+@@ -28,6 +28,7 @@ static DEFINE_RAW_SPINLOCK(irq_lock);
+ 
+ #define REG(offset, cpu)	(realtek_ictl_base[cpu] + offset)
```
you reccon we can avoid this somewhat nasty define? I suppose once we write a 'rtl-otto-irq' driver, that gets is address from the devicetree, this could somehow go away a that point ... It just looks horrendous, then again, most of our code here does ... meh

plappermaul replied:
Does anyone read "remove a nasty define" in the patch message?

oliv3r commented on target/linux/realtek/patches-5.10/319-irqchip-irq-realtek-rtl-fix-VPE-affinity.patch:
```diff
@@ -0,0 +1,145 @@
++{
++	u32 value;
++
++	value = readl(REG(RTL_ICTL_GIMR, cpu));
```
any reason to avoid the preferred ioread32 (other then consistency within the rest of the file?)

plappermaul replied:
Just stay consistent with the old syntax. Someone else can clean it up when patches are applied.

oliv3r replied:
That's a good enough reason for me.

oliv3r commented:
General question, with these fixes, we can drop `files-5.10/arch/mips/kernel/cevt-rtl9300.c` completely can't we? Even means we can drop the symbol `CONFIG_CEVT_RTL9300` right?

plappermaul commented:
> General question, with these fixes, we can drop `files-5.10/arch/mips/kernel/cevt-rtl9300.c` completely can't we? Even means we can drop the symbol `CONFIG_CEVT_RTL9300` right?

Not with this PR. But later after applying the timer PR #10573 10631 I provided. When activating it for the 930x we will finally drop the CEVT stuff.

oliv3r commented:
> > General question, with these fixes, we can drop `files-5.10/arch/mips/kernel/cevt-rtl9300.c` completely can't we? Even means we can drop the symbol `CONFIG_CEVT_RTL9300` right?
> 
> Not with this PR. But later after applying the timer PR #10631 I provided. When activating it for the 930x we will finally drop the CEVT stuff.

Ah fair enough; starting to mix up my MR's :p lots going on!

oliv3r commented:
Btw, I have no idea if this is a false positive (I would _not_ be supprised it to be false :p) but just an FYI.

```
git am /tmp/10573.patch                                                                                                                               2022-09-13 21:51:21 [0]
Applying: realtek: avoid wrong interrupt routing
.git/rebase-apply/patch:15: trailing whitespace.
 
.git/rebase-apply/patch:17: trailing whitespace.
 
.git/rebase-apply/patch:21: trailing whitespace.
 
.git/rebase-apply/patch:24: trailing whitespace.
 
.git/rebase-apply/patch:55: trailing whitespace.
 
warning: squelched 39 whitespace errors
warning: 43 lines applied after fixing whitespace errors.
Applying: realtek: use hotplug framework for setup during secondary CPU start
.git/rebase-apply/patch:19: space before tab in indent.
 	/* This is Malta specific: IPI,performance and timer interrupts */
.git/rebase-apply/patch:21: space before tab in indent.
 	if (mips_gic_present())
.git/rebase-apply/patch:22: space before tab in indent.
 		change_c0_status(ST0_IM, STATUSF_IP2 | STATUSF_IP3 |
.git/rebase-apply/patch:23: space before tab in indent.
 					 STATUSF_IP4 | STATUSF_IP5 |
.git/rebase-apply/patch:43: trailing whitespace.
 
warning: squelched 32 whitespace errors
warning: 32 lines applied after fixing whitespace errors.
```

plappermaul commented:
In my opinion this is a very central knot to be unleashed. From my point of view:

- No maintainer complaints about the first commit yet. -> seems to be ok.
- Lantiq has taken over the IRQ enabling from the second commit of this PR. -> seems to be ok.

merge?

plappermaul commented:
v3 is a even cleaner version. Would be helpful to get some feedback what we wait for here.

abajk commented on target/linux/realtek/files-5.10/arch/mips/rtl838x/prom.c:
```diff
@@ -193,15 +217,17 @@ void __init prom_init(void)
+	if (cpu_has_mipsmt) {
+		rtl_smp_ops = vsmp_smp_ops;
+		rtl_smp_ops.init_secondary = rtl_init_secondary;
+		register_smp_ops(&rtls_smp_ops);
```
Typo in rtls_smp_ops.

plappermaul replied:
yup

abajk commented on target/linux/realtek/files-5.10/arch/mips/rtl838x/prom.c:
```diff
@@ -29,10 +29,34 @@ extern const char __appended_dtb;
+	else
+		set_c0_status(ST0_IM);
+}
+
```
The #endif should be here.

plappermaul replied:
of corse

abajk commented on target/linux/realtek/files-5.10/arch/mips/rtl838x/prom.c:
```diff
@@ -29,10 +29,34 @@ extern const char __appended_dtb;
+#ifdef CONFIG_MIPS_MT_SMP
+extern const struct plat_smp_ops vsmp_smp_ops;
+
+static struct plat_smp_ops rtl_smp_ops;
```
You can reduce the scope of this variable and move it before the line:
`rtl_smp_ops = vsmp_smp_ops;`

plappermaul replied:
Done

abajk commented on target/linux/realtek/files-5.10/arch/mips/rtl838x/prom.c:
```diff
@@ -29,10 +29,34 @@ extern const char __appended_dtb;
+ * Malta if external GIC is available. Maybe we need this too.
+ */
+	if (mips_gic_present())
+		pr_warn("%s: GIC present. Maybe interrupt enabling required.\n", __func__);
```
RTL931x use a GIC or a custom interrupt controller?

plappermaul replied:
It is an Interaptiv and according to https://www.linux-mips.org/wiki/Aptiv there should be a GIC.

plappermaul committed:
realtek: use upstream recommendation for secondary CPU start

Currently we fix interrupts/timers for the secondary CPU by patching
vsmp_init_secondary(). Get a little bit more generic and use the
upstream recommended way instead. Additionally avoid a check around
register_cps_smp_ops() because it does that itself.

See https://lkml.org/lkml/2022/9/12/522

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul commented:
@abajk thanks for spotting my faults

svanheule commented:
Thanks, merged with:
bcb5d6b21ba5 realtek: avoid wrong interrupt routing
3f41360eb70c realtek: use upstream recommendation for secondary CPU start

I think the Realtek interrupt driver (and maybe MIPS CPU IRQ driver) will need other fixes upstream, but let's run with this until somebody has time to do that.

abajk commented on target/linux/realtek/files-5.10/arch/mips/rtl838x/prom.c:
```diff
@@ -193,15 +215,19 @@ void __init prom_init(void)
 #ifdef CONFIG_MIPS_MT_SMP
-	if (!register_vsmp_smp_ops()) {
+	if (cpu_has_mipsmt) {
+		struct plat_smp_ops rtl_smp_ops;
```
@plappermaul Does it work without the static keyword?
On lantiq target kernel then can't boot.

plappermaul replied:
I don't get the point. You recommended to reduce the scope of the variable. Is this wrong?

abajk replied:
After testing on the hardware, I think my recommendation was wrong. In fact, I don't know why it doesn't work after limiting the scope of the variable.
